lib/rspec/bash/stubbed_env.rb

```diff
@@ -2,78 +2,122 @@
-      attr_reader :dir
+      RUBY_STUB = :ruby_stub
+      BASH_STUB = :bash_stub
 
-      def initialize
-        @dir = Dir.mktmpdir
-        at_exit { cleanup }
+      def initialize(stub_type)
+        start_stub_server(stub_type)
       end
 
-      def cleanup
-        FileUtils.remove_entry_secure @dir if Pathname.new(@dir).exist?
+      def start_stub_server(stub_type)
+        tcp_server = create_tcp_server
+        stub_server = create_stub_server(stub_type)
+        stub_server.start(tcp_server)
       end
 
       def stub_command(command)
-        StubbedCommand.new(command, @dir)
+        check_if_command_is_allowed(command)
+        add_override_for_command(command)
+        create_stubbed_command(command)
       end
 
       def execute(command, env_vars = {})
-        full_command = get_wrapped_execution_with_function_overrides(
-          <<-multiline_script
-            source #{command}
-          multiline_script
-        )
-
-        Open3.capture3(env_vars, full_command)
+        script_runner = "source #{command}"
+        script_wrapper = wrap_script(script_runner)
+        execute_script(env_vars, script_wrapper)
       end
 
       def execute_function(script, command, env_vars = {})
-        full_command = get_wrapped_execution_with_function_overrides(
-          <<-multiline_script
-            source #{script}
-            #{command}
-          multiline_script
-        )
-
-        Open3.capture3(env_vars, full_command)
+        script_runner = "source #{script}\n#{command}"
+        script_wrapper = wrap_script(script_runner)
+        execute_script(env_vars, script_wrapper)
       end
 
       def execute_inline(command_string, env_vars = {})
-        temp_command_path = Dir::Tmpname.make_tmpname("#{@dir}/inline-", nil)
-        File.write(temp_command_path, command_string)
-        execute(temp_command_path, env_vars)
+        temp_command_path = create_file('inline-')
+        write_file(temp_command_path, command_string)
+        stdout, stderr, status = execute(temp_command_path, env_vars)
+        delete_file(temp_command_path)
+        [stdout, stderr, status]
       end
 
       private
 
-      def get_wrapped_execution_with_function_overrides(execution_snippet)
-        execution_binding_for_template = execution_snippet
-        function_override_path_binding_for_template = "#{@dir}/*_overrides.sh"
-        wrapped_error_path_binding_for_template = "#{@dir}/errors"
+      STUB_MARSHALLER_MAPPINGS = {
+        RUBY_STUB => RubyStubMarshaller,
+        BASH_STUB => BashStubMarshaller
+      }.freeze
+      STUB_SCRIPT_MAPPINGS = {
+        RUBY_STUB => RubyStubScript,
+        BASH_STUB => BashStubScript
+      }.freeze
+      DISALLOWED_COMMANDS = %w(/usr/bin/env bash readonly function).freeze
+
+      def create_tcp_server
+        tcp_server = TCPServer.new('localhost', 0)
+        @stub_server_port = tcp_server.addr[1]
+        tcp_server
+      end
+
+      def create_stub_server(stub_type)
```

Comment:
Instead of passing stub_type around, should we just use inheritance and have different classes depending if we're using the ruby stub vs. the bash stub?